dvc/repo/artifacts.py

```diff
@@ -50,18 +59,31 @@ def read(self) -> Dict[str, Dict[str, Artifact]]:
             dvcfile,
             dvcfile_artifacts,
         ) in self.repo.index._artifacts.items():  # pylint: disable=protected-access
-            # read the artifacts.yaml file if needed
-            if isinstance(dvcfile_artifacts, str):
-                dvcfile_artifacts = ArtifactsFile(
-                    self.repo,
-                    os.path.join(os.path.dirname(dvcfile), dvcfile_artifacts),
-                    verify=False,
-                ).load()
-            if not dvcfile_artifacts:
-                continue
             dvcyaml = relpath(dvcfile, self.repo.root_dir)
             artifacts[dvcyaml] = {}
             for name, value in dvcfile_artifacts.items():
                 check_name_format(name)
                 artifacts[dvcyaml][name] = Artifact(**value)
         return artifacts
+
+    def add(self, name: str, artifact: Artifact, dvcfile: Optional[str] = None):
+        # this doesn't update it "in place": self.read() won't return the updated value
+        if not dvcfile or dvcfile[-len("dvc.yaml") :] != "dvc.yaml":
+            dvcfile = os.path.join(dvcfile or "", "dvc.yaml")
+        dvcfile_abspath = os.path.join(self.repo.root_dir, dvcfile)
+        Path(os.path.dirname(dvcfile_abspath)).mkdir(exist_ok=True, parents=True)
+        _update_project_file(
+            ProjectFile(self.repo, dvcfile_abspath), name, artifact.to_dict()
+        )
+        return dvcfile_abspath
```

Comment:
This is for Studio. Once you add an artifact, you need to git add & commit changes. Since DVC should know where this is going to be written, I think it's valid to return that path. Can return both `Artifact` and path. WDYT @omesser?

Reply:
Alternative way is to calculate this on Studio side. Since we pass `dvc.yaml` path there, I can simply use it to stage changes. It can make more sense from DVC perspective, so that can be a better way. Didn't implement it first since I thought: if `dvc.yaml` path calculation changes, we're going to need to update Studio BE to address that, so wanted to keep responsibility on DVC side here.